Errata including Grammar and term references

gkellogg opened the pull request:
* Update grammar rules on whitespace and comments described in erratum 24.
  Note that this does not includee the explicit `WS` terminal and updates to the EBNF due to LL(1) problems introduced.
* Term references.

I think the white space rules in the grammar section are adequate to know how to properly parse the data, and most RDF EBNF grammars allow white space the same way.

Updating the EBNF to be explicit, and keep it context free, probably require the introduction of new rules to remove the conflicts that would otherwise be introduced.

Also, note that N-Triples is already described as a subset of Turtle which more accurately describes its history. Changing this so that Turtle is described as an extension of N-Triples does not add anything further, so I believe that [Erratum 17](https://www.w3.org/2001/sw/wiki/RDF1.1_Errata#erratum_17) does not help make things clearer.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-n-triples/pull/6.html" title="Last updated on Feb 13, 2023, 7:03 PM UTC (f4fd25b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-n-triples/6/a7407b6...f4fd25b.html" title="Last updated on Feb 13, 2023, 7:03 PM UTC (f4fd25b)">Diff</a>

afs commented on spec/index.html:
```diff
@@ -280,7 +308,8 @@ <h3>Other Media Types</h3>
     Unicode code points only in the range U+0 to U+10FFFF inclusive are allowed.</p>
-    <p>White space (tab <code>U+0009</code> or space <code>U+0020</code>) is used to separate two terminals which would otherwise be (mis-)recognized as one terminal. White space is significant in the production <a href="#grammar-production-STRING_LITERAL_QUOTE">STRING_LITERAL_QUOTE</a>.</p>
+    <p>White space (tab U+0009 or space U+0020) is allowed but not required between any two terminals.
+      White space is significant in the production <a href="#grammar-production-STRING_LITERAL_QUOTE">STRING_LITERAL_QUOTE</a>.</p>
```
I don't know if here is the best to fit it in but it is worth nothing that white space (horizontal) in NT is not the same as white space in Turtle.

This does not need addressing here - there are other white space details to sort out  -- #7.

pfps reviewed (commented):
I have no comments, but this is not my area of expertise.

gkellogg commented:
@pfps I tagged you because the grammar erratum came from you. No need to comment further.

pfps commented on spec/index.html:
```diff
@@ -104,24 +108,38 @@ <h2>Introduction</h2>
+      of <a href="#simple-triples">simple triples</a>,
+      one for each line.
+      Lines consisteng entirely of white space  (spaces <code>U+0020</code> or tabs <code>U+0009</code>)
+      and/or comments are permitted before or after each triple.</p>
```
Why not just "lines consisting entirely of white space ... and/or comments are permitted"?

afs replied:
Calling out comments here implies they on a line of their own which isn't the case. It may be what was intended (and is a good idea) but section 3 (non-normative) is different -- "after a triple production".

We could argue it (comments on their own line) was the intent.

gkellogg replied:
Example 1 explicitly shows a comment after the tailing `.`, so the intent is clearly not that comments be on their own lines. I'll work on the wording.

afs replied:
You're right. Shame.

pfps commented on spec/index.html:
```diff
@@ -104,24 +108,38 @@ <h2>Introduction</h2>
+        (<a data-cite="RDF12-CONCEPTS#dfn-subject">subject</a>,
+        <a data-cite="RDF12-CONCEPTS#dfn-predicate">predicate</a>,
+        <a data-cite="RDF12-CONCEPTS#dfn-object">object</a>) terms,
+        and terminated by '<code>.</code>' after each triple.
```
Ditch the "after each triple"

pfps commented on spec/index.html:
```diff
@@ -104,24 +108,38 @@ <h2>Introduction</h2>
+        <a data-cite="RDF12-CONCEPTS#dfn-predicate">predicate</a>,
+        <a data-cite="RDF12-CONCEPTS#dfn-object">object</a>) terms,
+        and terminated by '<code>.</code>' after each triple.
+        White space MAY occur beteween elements of this sequence.
```
tyop

gkellogg commented on spec/index.html:
```diff
@@ -295,6 +324,12 @@ <h2>Grammar</h2>
+      suggests adding a `WS` terminal and making the use of white space explicit in the EBNF.
+      However, this leads to a grammar which is not LL(1) due to first/follow conflicts.
+    </p>
+
```
There is no hard and fast rule that W3C grammars be context-free, but generally they are so would rather keep that way. I believe a PEG parser wouldn't have the same issue.

With the caveat that this is going into too much detail, but for the record, with some minor re-writing for associativity, the grammar from the erratum is:

```
[1]  ntriplesDoc        ::= triple? (EOL triple)* END
[2]  triple             ::= WS? subject WS? predicate WS? object WS? '.'
[3]  subject            ::= IRIREF | BLANK_NODE_LABEL
[4]  predicate          ::= IRIREF 
[5]  object             ::= IRIREF | BLANK_NODE_LABEL | literal
[6]  literal            ::= STRING_LITERAL_QUOTE (WS? (('^^' WS? IRIREF) | LANGTAG))?

@terminals

[144s] LANGTAG          ::= "@" [a-zA-Z]+ ( "-" [a-zA-Z0-9]+ )* 

[7]  EOL                ::= [#xD#xA]+
[7a] END                ::= 	EOL? WS? ('#x22' [^#xD#xA]* )?

[8] IRIREF ::=  '<' ([^#x00-#x20<>"{}|^`\] | UCHAR)* '>'
[9] STRING_LITERAL_QUOTE ::= '"' ( [^#x22#x5C#xA#xD] | ECHAR | UCHAR )* '"' 

[141s] BLANK_NODE_LABEL ::= '_:' ( PN_CHARS_U | [0-9] ) ((PN_CHARS|'.')* PN_CHARS)?
[10] UCHAR ::= ( "\u" HEX HEX HEX HEX ) 
 | ( "\U" HEX HEX HEX HEX HEX HEX HEX HEX ) 
[153s] ECHAR ::= "\" [tbnrf"'] 
[157s] PN_CHARS_BASE    ::= [A-Z] 
                        | [a-z] 
                        | [#x00C0-#x00D6] 
                        | [#x00D8-#x00F6] 
                        | [#x00F8-#x02FF] 
                        | [#x0370-#x037D] 
                        | [#x037F-#x1FFF] 
                        | [#x200C-#x200D] 
                        | [#x2070-#x218F] 
                        | [#x2C00-#x2FEF] 
                        | [#x3001-#xD7FF] 
                        | [#xF900-#xFDCF] 
                        | [#xFDF0-#xFFFD] 
                        | [#x10000-#xEFFFF] 
[158s] PN_CHARS_U       ::=  PN_CHARS_BASE 
                        | '_' 
                        | ':'
[160s] PN_CHARS         ::= PN_CHARS_U 
                        | "-" 
                        | [0-9] 
                        | #x00B7 
                        | [#x0300-#x036F] 
                        | [#x203F-#x2040] 
[162s] HEX              ::= [0-9] | [A-F] | [a-f]
[163] WS                ::=	[#x9#x20]+
```

Parsing this as LL(1) (using my own [`ebnf`](https://github.com/dryruby/ebnf) implementation) results in the following error from my EBNF parser:

> `[1]First/Follow Conflict: :WS is both first and follow of _literal_1`

Written in an intermediate S-Expression, the `literal` production looks like the following:

```
 (rule literal "6"
  (seq STRING_LITERAL_QUOTE
   (opt
    (seq
     (opt WS)
    (alt
     (seq
      "^^"
      (opt WS)
      IRIREF)
     LANGTAG)))))
```

Flattening to  straight BNF results in:

```
[6]    literal              ::= STRING_LITERAL_QUOTE _literal_1
[6.1]  _literal_1           ::= _empty | _literal_2
[6.2]  _literal_2           ::= _literal_3 _literal_4
[6.3]  _literal_3           ::= _empty | WS
[6.4]  _literal_4           ::= _literal_5 | LANGTAG
[6.5]  _literal_5           ::= "^^" _literal_6 IRIREF
[6.6]  _literal_6           ::= _empty | WS
```

So, the error is saying that there is a first/follow conflict with `[6.1]  _literal_1           ::= _empty | _literal_2`. This is because both `_literal_1` can start with `_empty` and `_literal_3` can start with `_empty`, so there is no unique production when the parser sees `_empty`. Of course, there I'm sure there are other ways to re-write the production to avoid the conflict, but as @afs points out, the original is more readable and the parser can be given simple rules to skip white space when tokenizing the input.

gkellogg commented:
@afs, a fair number of changes, with text borrowed from Turtle and TriG, where appropriate. Also addressed your comment on producing the c14n version. This should also address #7. I'll refrain on doing parallel updates to https://github.com/w3c/rdf-n-quads/pull/11 until we're settled here.

The "blank lines" discussion is a bit hand-wavy, but should be adequate.

afs commented on spec/index.html:
```diff
@@ -279,13 +339,41 @@ <h3>Other Media Types</h3>
-      up-to, and excluding, the end of line (<code>EOL</code>), 
-      or end of file if there is no end of line after the comment
-      marker. Comments are treated as white space.</p>
+      Unicode code points only in the range U+0 to U+10FFFF inclusive are allowed.</p>
```
```suggestion
      </p>
```
Unicode only goes up to U+10FFFF. Bu saying "Unicode code points"  (or "Unicode string") is already excluding anything else and any 21 bit integers that are not valid codepoints.

afs commented on spec/index.html:
```diff
@@ -279,13 +339,41 @@ <h3>Other Media Types</h3>
+    <section id="sec-grammar-comments">
+      <h3>Comments</h3>
+
+      <p>Comments in N-Triples take the form of '<code>#</code>',
```
```suggestion
      <p>Comments in N-Triples start at '<code>#</code>'
```

afs commented on spec/index.html:
```diff
@@ -279,13 +339,41 @@ <h3>Other Media Types</h3>
+    <section id="sec-grammar-ws">
+      <h3>White Space</h3>
+
+      <p>White space (tab U+0009 or space U+0020) is allowed, but not required, before or after terminals.
```
```suggestion
      <p>White space (tab U+0009 or space U+0020) is allowed outside of terminals.
```

afs commented on spec/index.html:
```diff
@@ -279,13 +339,41 @@ <h3>Other Media Types</h3>
+
+      <p>White space is significant in the production <a href="#grammar-production-STRING_LITERAL_QUOTE">STRING_LITERAL_QUOTE</a>.</p>
+
+      <p>Blank lines, consisting of only white space and/or comments,
```
```suggestion
      <p>A blank line, consisting of only white space and/or a comment,
```

afs commented on spec/index.html:
```diff
@@ -279,13 +339,41 @@ <h3>Other Media Types</h3>
+      <h3>White Space</h3>
+
+      <p>White space (tab U+0009 or space U+0020) is allowed, but not required, before or after terminals.
+        Rule names below in capitals indicate where white space is significant;
+        these form a possible choice of terminals for constructing an N-Triples parser.</p>
```
```suggestion
        Rule names below in capitals indicate where white space is significant.
</p>
```
Drop the "possible choice". The grammar actually says they are terminals.

afs commented on spec/index.html:
```diff
@@ -279,13 +339,41 @@ <h3>Other Media Types</h3>
+      <p>White space is significant in the production <a href="#grammar-production-STRING_LITERAL_QUOTE">STRING_LITERAL_QUOTE</a>.</p>
+
+      <p>Blank lines, consisting of only white space and/or comments,
+        MAY appear wherever a <code><a href="#grammar-production-triple">triple</a></code> production is allowed,
```
Doesn't have to be an RFC "MAY".

gkellogg replied:
```suggestion
        may appear wherever a <code><a href="#grammar-production-triple">triple</a></code> production is allowed,
```

afs commented on spec/index.html:
```diff
@@ -295,6 +383,12 @@ <h2>Grammar</h2>
+    <p class="ednote">
+      <a href="https://www.w3.org/2001/sw/wiki/RDF1.1_Errata#erratum_24">Erratum 24</a>
+      suggests adding a `WS` terminal and making the use of white space explicit in the EBNF.
+      However, this leads to a grammar which is not LL(1) due to first/follow conflicts.
```
Explain based on legibility (if this ednote is needed at all).

gkellogg replied:
The ednote was intended to motivate any discussion, but it seems that's been revolved now, so can simply be removed.

afs commented:
> I'll refrain on doing parallel updates to [w3c/rdf-n-quads#11](https://github.com/w3c/rdf-n-quads/pull/11) until we're settled here.

👍🏻 I added an N-Quads issue to tick off later.

afs commented on spec/index.html:
```diff
@@ -145,17 +171,32 @@ <h3>IRIs</h3>
+        <code title="CARRIAGE RETURN"><sub>CR</sub></code> (<span class="codepoint">U+000D</span>)
+        except in their escaped forms.
+        In addition '<code>\</code>' (<span class="codepoint">U+005C</span>)
+        may not appear in any quoted literal except as part of an escape sequence.</p>
```
The Turtle text is imprecise on escaping : w3c/rdf-turtle#5

Add:
```suggestion
        may not appear in any quoted literal except as part of an escape sequence
        and a <code>"</code> (<span class="codepoint">U+0022</span>) character
        can only be included in a quote literal using an escape sequence.
        </p>
```
It's not necessary but if one spec (Turtle) mentions it, the others ought to.

gkellogg commented:
Fixes #7.

afs commented on spec/index.html:
```diff
@@ -233,6 +289,13 @@ <h2>A Canonical form of N-Triples</h2>
         allowed directly in
         <a href="#grammar-production-STRING_LITERAL_QUOTE">STRING_LITERAL_QUOTE</a>. </li>
     </ul>
```
Yes - I have been concentrating on n-quads C14N at FPWD as that is the immediate downstream requirement.

gkellogg commented on spec/index.html:
```diff
@@ -89,12 +89,17 @@ <h2>Introduction</h2>
+      of an <a data-cite="RDF12-CONCEPTS#dfn-rdf-triple">RDF Triple</a>.
       These may be separated by white space (spaces <code>U+0020</code> or tabs <code>U+0009</code>).
       This sequence is terminated by a '<code>.</code>'
+      (optionaly proceded by white space and/or a comment),
```
I'll correct it in the next set of commits.
